### ISSUE TITLE
Console not pairing since update to 12.0.0

### ISSUE THREAD

**Issue by thxomas** [Poohl/joycontrol#3]
I had a working setup on rpi3B with some issues, and after dist-upgrade (bad idea), I cannot connect anymore from the Grip/Order menu.

Joycontrol is indefinitely waiting for Switch to connect :
- hcidump shows request from the Switch, but after some exchanges, it terminates after passkey sent (HCI Auth Failure, Remote Terminated)
- I found that the Device Class is not properly set (even if the command returns correct in python, hciconfig shows 0x000000. Setting manually to 0x002508 is reset to 0x000000 when re-running joycontrol). That may explain the Authentication Failure.
- so now I have doubts about the bluez5 (don't know my previous version), but the SDP seems correctly set (Nintendo Wireless Gamepad with L2CAP+HIDP) However, the AuthenticationRequired=False may not be honoured, another possible explanation to the Auth Failure

[EDIT: unrelated issue fixed.]

I'm trying to fix this before being able to troubleshoot amiibo_writing again. :)

**Comment by thxomas** [comment]
Update : fixed Device Class setting by doing it AFTER hid.discoverable() in server.py

However, still no pairing from the switch, here is the end of hcidump : 
```
> HCI Event: User Confirmation Request (0x33) plen 10
    bdaddr 94:58:CB:XX:XX:XX passkey 185519
< HCI Command: User Confirmation Request Negative Reply (0x01|0x002d) plen 6
    bdaddr 94:58:CB:XX:XX:XX
> HCI Event: Command Complete (0x0e) plen 10
    User Confirmation Request Negative Reply (0x01|0x002d) ncmd 1
    status 0x00 bdaddr 94:58:CB:XX:XX:XX
> HCI Event: Simple Pairing Complete (0x36) plen 7
    status 0x05 bdaddr 94:58:CB:XX:XX:XX
    Error: Authentication Failure
> HCI Event: Disconn Complete (0x05) plen 4
    status 0x00 handle 12 reason 0x13
    Reason: Remote User Terminated Connection
```
The rpi seems to reject the passkey although the AuthenticationRequired=False flag. So nothing happens at python level.
Investigating if this is a new default behaviour of the bluez version.

T

**Comment by thxomas** [comment]
OK, what's the practice on git ? Rename an issue or create a new one ?
I didn't notice the update to 12.0.0... 🤔 so this issue is merely caused by the console, not linux...

**Comment by thxomas** [comment]
Went a step further in the pairing process, by setting a bt-agent to accept the passkey.
Next step is failing after " Link Key Notification"

**Comment by thxomas** [comment]
If anyone interested : After gathering info from various projects, I managed to get the amiibo_writing branch work on raspberry pi 3, latest raspbian buster.
Amiibo read/write is not yet functional, but initial pairing works and then using reconnect gives a stable connection to the switch (I was able to launch BOTW from Home menu, load a save, and trigger amiibo reading, but was greeted by "Connect a controller that can read amiibos..."). 

Please be aware it is still manual, and not ready for easy sharing. Steps to reproduce : 
- **Change RPi BT MAC address to get in Nintendo OUI range** : I just changed the first 3 byte of the original to 94:58:CB
`hcitool cmd 0x3f 0x001 0xbe 0x13 0x37 0xCB 0x58 0x94` (MAC bytes in reverse order
  or
use the `bdaddr` tool 
- **Disable extra SDP Profiles** by setting ExecStart command in systemd /lib/systemd/system/bluetooth.service to :
`ExecStart=/usr/lib/bluetooth/bluetoothd -C -P sap,input,avrcp`
- **Force default device class** (I noticed bluez tends to set this value at unwanted times)  in /etc/bluetooth/main.conf : 
`Class = 0x002508`
- **Set default hostname** in /etc/machine-info : 
`PRETTY_HOSTNAME=Pro Controller`
- **Configure an Agent** to negotiate pairing security : I used the test/simple-agent provided in bluez source, and slightly modified it to accept anything without blocking prompt. Launch it in the background.
`./simple-agent &`
- Modify `server.py` by swapping hid.discoverable() and set_class()

```
        # start advertising
        hid.discoverable()

        # set the device class to "Gamepad/joystick"
        await hid.set_class()
```


When this is done : try pairing..
`sudo ./run-controller.py PRO_CONTROLLER`
and go to Grip Screen, connection works ! However if you send 'a' command, the menu exits, but "ERROR - Connection lost"

Now if you reconnect with -r option, the connection keeps ON, until trying NFC read/write.


Of course, these steps should be in some way integrated to the project, but at least the POC is done.
Another drawback, is that the RPi is somewhat dedicated only to emulating a joycon, and trying any other BT service would break it..

Thanks to @Brikwerk from NXBT project for having found the missing information to reach this state

ref. [https://github.com/Brikwerk/nxbt/issues/18](https://github.com/Brikwerk/nxbt/issues/18)

I need amiibos back  ! 💪 
T.

**Comment by thxomas** [comment]
> I am trying your solution on my HP laptop with intel bluetooth on ubuntu and I followed your steps but I got stuck
> this file is missing /etc/machine-info

This file is specific to the RPi I think. this is the way bluez gets the hostname for bluetooth. On Ubuntu, you may check /etc/bluetooth/main.conf for the Name setting (top of file, just before "Class")

> and where do I get test/simple-agent ? and what is the modification I need to change ?

The file is an example found in the bluez source package. I attached my (dirty) modified version to this comment (rename .py or remove extension) Unlike joycontrol it runs on python2, you may need to `apt install python-bluez`

[simple-agent.txt](https://github.com/Poohl/joycontrol/files/6326493/simple-agent.txt)

Does the MAC Address change have an effect on the Intel chipset ?
Good luck
T.

**Comment by thxomas** [comment]
Run joycontrol, then, in another shell session, check your controller status by comparing to this : 

```
/home/pi# bluetoothctl show

Controller 94:58:CB:96:BC:79 (public)           #MAC Address must be in Nintendo OUI range
        Name: Pro Controller                    #Name and Alias like a true controller
        Alias: Pro Controller
        Class: 0x00002508                       # Set by joycontrol
        Powered: yes
        Discoverable: yes
        Pairable: yes
        UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)   # Default profile (accepted by switch)
        UUID: Human Interface Device... (00001124-0000-1000-8000-00805f9b34fb)   # Gamepad profile set by joycontrol
        UUID: PnP Information           (00001200-0000-1000-8000-00805f9b34fb)   # Default profile (accepted by switch)
        UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)   # Default profile (accepted by switch)
        Modalias: usb:v1D6Bp0246d0532
        Discovering: no

```

The simple-agent must be running detached, or in another session, it will auto accept pairing. Once paired, you don't need it to be running to reconnect with `-r` option.

T

**Comment by JaredEzz** [comment]
Thanks to @thxomas for providing steps to troubleshoot.

**1. Change RPi BT MAC address to get in Nintendo OUI range**
I installed the libbluetooth-dev & bdaddr tools to help with this, follow instructions [here](https://scribles.net/changing-a-bluetooth-device-address-on-raspberry-pi/).
I'm using a raspberry pi 2 with a usb bluetooth connector, this was the info before the change:
```
Manufacturer:   Cambridge Silicon Radio (10)
Device address: 00:1A:7D:DA:71:13
```
and after the change:
```
Manufacturer:   Cambridge Silicon Radio (10)
Device address: 00:1A:7D:DA:71:13
New BD address: 94:58:CB:DA:71:13

Address changed - Reset device manually
```
**2. Disable extra SDP Profiles**
I changed ExecStart in /lib/systemd/system/bluetooth.service from 
`ExecStart=/usr/lib/bluetooth/bluetoothd --noplugin=input` to 
`ExecStart=/usr/lib/bluetooth/bluetoothd -C -P sap,input,avrcp`

**3. Force default device class** 
I changed Class in /etc/bluetooth/main.conf (under the [General] section at the top) from 
`#Class = 0x000100` to
`Class = 0x002508`

**4. Set default hostname in /etc/machine-info**
_I also did not find the machine-info file on my raspberry pi, could be because of the linux version, rpi version, etc._
I changed Name in /etc/bluetooth/main.conf from
`Name = BlueZ` to
`Name = Pro Controller`

**5. Configure an Agent**
I copied the contents of the prived `simple-agent.txt` into `simple-agent` on my rpi, made it executable and ran it, and got some missing packages errors. I installed the recommended package `python-bluez`, as well as `python-gobject`, `python-dbus` because of import errors, and I got the following error:
```
ERROR:dbus.proxies:Introspect error on :1.5:/org/bluez: dbus.exceptions.DBusException: org.freedesktop.DBus.Error.AccessDenied: Rejected send message, 2 matched rules; type="method_call", sender=":1.39" (uid=1001 pid=1087 comm="/usr/bin/python ./simple-agent ") interface="org.freedesktop.DBus.Introspectable" member="Introspect" error name="(unset)" requested_reply="0" destination=":1.5" (uid=0 pid=280 comm="/usr/lib/bluetooth/bluetoothd --noplugin=input ")
Traceback (most recent call last):
  File "./simple-agent", line 161, in <module>
    manager.RegisterAgent(path, capability)
  File "/usr/lib/python2.7/dist-packages/dbus/proxies.py", line 70, in __call__
    return self._proxy_method(*args, **keywords)
  File "/usr/lib/python2.7/dist-packages/dbus/proxies.py", line 145, in __call__
    **keywords)
  File "/usr/lib/python2.7/dist-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.AccessDenied: Rejected send message, 2 matched rules; type="method_call", sender=":1.39" (uid=1001 pid=1087 comm="/usr/bin/python ./simple-agent ") interface="org.bluez.AgentManager1" member="RegisterAgent" error name="(unset)" requested_reply="0" destination=":1.5" (uid=0 pid=280 comm="/usr/lib/bluetooth/bluetoothd --noplugin=input ")
```

Moving on to try the rest without the agent but wanted to post in case there's a simple fix for this error

**Comment by thxomas** [comment]
Try sudo or running the agent as root

**Comment by JaredEzz** [comment]
**6. Modify server.py by swapping hid.discoverable() and set_class()**
did this ^^ self-explanatory

**FINAL** 
ran `sudo ./run-controller.py PRO_CONTROLLER` and it ran as expected. Paired, connected, then disconnected after a command or two. 
Then ran the command with the -r flag and everything works as expected! Turns out I didn't need the agent. Thank you so much @thxomas

**Comment by thxomas** [comment]
For more insights if you want, find and install bluez-utils, to get tools liike btlon or btmgmt. They give good clues to the cause of problems (Switch not connecting at all, Authentication errors, etc...)

**Comment by JaredEzz** [comment]
I spoke too soon. It disconnected after a minute or so

**Comment by thxomas** [comment]
Good and bad news !
Looks like we will have to deal with the command rate spottef by NXBT 😭

Maybe @Poohl  will look into it ?

**Comment by JaredEzz** [comment]
ran `bluetoothctl show` and it looks like the Alias is properly set but the name and class did not set correctly
![image](https://user-images.githubusercontent.com/13698465/115058430-7cc35a00-9ea2-11eb-90ba-a6b678645986.png)

**Comment by JaredEzz** [comment]
I manually added a `machine-info` file in /etc and added the PRETTY_HOSTNAME to it. That fixed the Name issue. Connection is still lost after a few commands however. Looking for a solution to the Class problem.

These are the relevant lines in my server.py
```
        # start advertising
        hid.discoverable()

        # setting bluetooth adapter name and class to the device we wish to emulate
        await hid.set_name(protocol.controller.device_name())
        await hid.set_class()
```

and again, in my /etc/bluetooth/main.conf
`Class = 0x002508`

@thxomas any thoughts?

**Comment by JaredEzz** [comment]
`sudo hciconfig hciX class 0x002508` changes the class, but the connection still drops after a minute or two

![image](https://user-images.githubusercontent.com/13698465/115061182-13dde100-9ea6-11eb-921c-d9124d6a3c2f.png)

**Comment by Poohl** [comment]
Wait, wait. Sorry for beeing late didn't get notified of the name change.

So do I get this right: @thxomas got it working again just by messing with the Bluetooth pairing process and not modifying the input frequency (or to much inside joycontrol in general)?

Since NXBT's solution is to limit the speed to 15Hz and 1hz before pairing, that can be done aswell but I highly doubt that this is how it should be done. I'll check old captures to check for something we might be doing wrong, @Brickwerk speculated [here](https://github.com/Brikwerk/nxbt/issues/18#issuecomment-818141802) that it might be to do with the timing byte or some timing in general.

I updated my switch while I was away too, damm it.

**Comment by JaredEzz** [comment]
@thxomas figured out how to pair it again. I can pair and connect again with -r but the connection only lives as long as I stay in the switch menus. Entering a game causes the connection to fail

**Comment by JaredEzz** [comment]
I noticed that the list of UUIDs is different when connecting the first time and reconnecting. The Human Interface Device UUID isn't present when I `bluetoothctl show` after reconnecting

**Comment by JaredEzz** [comment]
Any ideas on how I can apply the sdp_record_hid.xml manually? Not having any luck googling. I'm trying to do it with sdptool

**Comment by thxomas** [comment]
> @thxomas figured out how to pair it again. I can pair and connect again with -r but the connection only lives as long as I stay in the switch menus. Entering a game causes the connection to fail

Entering Zelda:BOTW does not disconnect, I'm so lucky :)

> I noticed that the list of UUIDs is different when connecting the first time and reconnecting. The Human Interface Device UUID isn't present when I bluetoothctl show after reconnecting
>Any ideas on how I can apply the sdp_record_hid.xml manually? Not having any luck googling. I'm trying to do it with sdptool.

I think no need to. SDP is intended for discovery only, once paired, the Switch will remember the UUIDs and PSM to connect to.

**Comment by JaredEzz** [comment]
OK, well Pokemon Sword/Shield disconnects :( probably the frequency issue. Just for fun, I connected to Zelda:BOTW and it works fine. I'll test some other games rn too.

**Comment by thxomas** [comment]
> So do I get this right: @thxomas got it working again just by messing with the Bluetooth pairing process and not modifying the input frequency (or to much inside joycontrol in general)? 

Yes sir ! The new pairing has additional security (PIN&Encryption) and the console does not tolerate much deviation from a true controller (Name, SDP profiles, ...). But I focused on BotW and apparently got lucky. I did not tinker with any core joycontrol logic as I don't get it yet. 
One paired, data exchange and controller init work well.
> 
> Since NXBT's solution is to limit the speed to 15Hz and 1hz before pairing, that can be done aswell but I highly doubt that this is how it should be done. I'll check old captures to check for something we might be doing wrong, @Brickwerk speculated [here](https://github.com/Brikwerk/nxbt/issues/18#issuecomment-818141802) that it might be to do with the timing byte or some timing in general.

I'll try to help you in the timing byte analysis when I understand it ;)

**Comment by Poohl** [comment]
@thxomas First I'm trying to get anything back to work. I followed @JaredEzz steps but my bluetooth adapter isn't complying. I have a raspi 4, but apparently these also have issues with that on top of the incredibly spotty wifi/bluetooth chip on there. Will take some time.

What I found about the timing byte: The switch just increments it by one each message, the joycon (right joycon) increments it by 1 every 0.005 seconds, but increments it a lot more on input mode changes or some MCU commands. Since the joycons run at 60hz, that means incrementing it by 3 every report seems sensible.

**Comment by Poohl** [comment]
@pokemon-bot Same problem here. If you find one that works please tell me which one (brand, model/type) because if my raspi won't cooperate either I gotta buy one :(

For the record:
* random Qualcomm-card inside my Acer laptop doesn't work
* Gigabyte PCI-E with Intel chipset never worked

**Comment by Poohl** [comment]
nope, haven't tried. Does it work by plugging into the joycon-rails?

All I've tried is [USB](https://github.com/Poohl/gadfetfs-Switch-Fightstick), but then found this and abandoned any attempts to get the PRO-Controller to talk via USB.

**Comment by thxomas** [comment]
Back to start :( I rebooted the pi, did my steps again, checked everything was OK, tried to pair -> NOK.
The pairing stops again after sending 'Link Key Notification'
Been struggling for 1 hour on bringing it back to work... 😠

**Comment by Poohl** [comment]
@thxomas Have you tried deleting `/var/lib/bluetooth`? I've had issues with linux just assuming some key and the switch trying to negotiate one. Or maybe delete/unregister the controller on the switch.

**Comment by JaredEzz** [comment]
@pokemon-bot yeah the arduino works, you can pm me on discord about it jaredezz#1338 but let's keep this thread to bluetooth only. I'd like to figure it out

**Comment by Poohl** [comment]
@JaredEzz What BT dongle are you using? the only thing with a Cambridge Silicon Radio chip insinde I just found still available is something by ZEXMTE.

**Comment by JaredEzz** [comment]
Yep, I'm using the ZEXMTE one. Here's a link: https://amzn.to/3suFokx

**Comment by Poohl** [comment]
@JaredEzz @pokemon-bot I figured out how to change the MAC on a raspi 4B, it should also work on a 3B+ as they use same chip+diver combo. See [my question](https://raspberrypi.stackexchange.com/questions/124107/pi-4-change-bluetooth-mac-adress/124117#124117) or TLDR:
```
bdaddr -i hci0 01:23:45:67:89:ab
hciconfig hci0 reset
systemctl restart bluetooth.service
```

**Comment by JaredEzz** [comment]
@Poohl Great! Good luck on the rest of it. Let me know how it goes? Really want to be able to run macros again

**Comment by Poohl** [comment]
New branch with patches is live. see [V12_fixes](https://github.com/Poohl/joycontrol/tree/V12_fixes)

Includes a script to adjust the BTADDR (for RPI4B & RPI3B+), the fix for the BT-Class and some Doc on how to get it up and running with the reconnect workaround. Pairing still just disconnects afterwards, am working on it.

Big thanks to @thxomas and @JaredEzz for figuring most of this out.

**Comment by Poohl** [comment]
Ok, very annoying find:

The switch will happily connect with only the alias beeing set, but once you are paired it will refuse to re-pair with the same PREATTY_NAME ever again. My fix is to just rotate the pretty name by going through the alphabet (`PRETTY_NAME='Pro Controllera'`)

**Comment by JaredEzz** [comment]
I never had a problem re-pairing as long as I removed the switch from my paired devices on the RPi first

**Comment by JaredEzz** [comment]
I hope you figure out how to get it to stay connected! Let me know, and what games you're testing it on

**Comment by Poohl** [comment]
@JaredEzz thanks, it was the paired issue. I always just assumed BlueZ was glitchy again and purged it.

FYI: I'm currently only testing in the menus (want to get the pairing to not die after leaving).

Games I test NFC on are BOTW, Splatoon2 and the System Settings but currently not working on that. 

If anyone wants to help with the NFC mechanim: take a look at [pipe-stub](https://github.com/Poohl/pipe-stub/tree/joycon-nfc) where I try to get a simple Finite-state machine to make a real joycon write to a real nfc tag, because we still have no Idea what happens _after_ the write is done. The Joycon reports some kind of processing for a while and then???

**Comment by Poohl** [comment]
Hi, I found a very subtle difference between the simulated and the real switch's regular input talk.

When comparing the real deal:
```
0000   a1 30 11 8e 40 00 00 00 00 00 c9 58 74 0c 0b f9
0010   25 f5 b4 f6 0e 97 00 e4 d6 3f fc f8 1f f5 c2 f6
0020   35 ff 27 07 e0 fd 00 f9 12 f5 d1 f6 05 10 fa 77
0030   1d 24
```
to the simulated one:
```
0000   a1 30 4e 8e 00 00 00 00 08 80 00 08 80 80
```
One can notice that 2/3 of the packet are missing.... am I blind? How does this even work at all? I'm very certain that this is the real deal, both were captured using wireshark and the driver-headers confirm this to be the actual length on wire.

Is the Switches implementation _that_ bad that it doesn't notice all the 6axis data to be _missing_?

I'm sorry if this is a very obvious/stupid question, I honestly never looked at the raw data in regular packets before (always used `nxbt.type != 0x30`).

**Comment by MadeSimpleWorkshop** [comment]
I'm able to connect with the above changes to "mart1nro/joycontrol"

After it sending a command 'b" it disconnects I tried the -r and get:
`
$ sudo python3 run_controller_cli.py -r 01:23:45:67:89:AB  PRO_CONTROLLER 

Traceback (most recent call last):
  File "run_controller_cli.py", line 329, in <module>
    _main(args)
  File "/usr/lib/python3.7/asyncio/base_events.py", line 584, in run_until_complete
    return future.result()
  File "run_controller_cli.py", line 287, in _main
    device_id=args.device_id)
  File "/home/pi/Development/joycontrol/joycontrol/server.py", line 115, in create_hid_server
    client_ctl.connect((reconnect_bt_addr, ctl_psm))
OSError: [Errno 113] No route to host

`

Which address to use after -r for reconnecting?  I tried sudo python3 run_controller_cli.py -r 01:23:45:67:89:AB  PRO_CONTROLLER 

`
 $ bluetoothctl show

Controller 01:23:45:67:89:AB (public)
	Name: Pro Controller
	Alias: Pro Controller
	Class: 0x001c0508
	Powered: yes
	Discoverable: no
	Pairable: no
	UUID: Headset AG                (00001112-0000-1000-8000-00805f9b34fb)
	UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
	UUID: OBEX File Transfer        (00001106-0000-1000-8000-00805f9b34fb)
	UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
	UUID: OBEX Object Push          (00001105-0000-1000-8000-00805f9b34fb)
	UUID: PnP Information           (00001200-0000-1000-8000-00805f9b34fb)
	UUID: IrMC Sync                 (00001104-0000-1000-8000-00805f9b34fb)
	UUID: Audio Source              (0000110a-0000-1000-8000-00805f9b34fb)
	UUID: Audio Sink                (0000110b-0000-1000-8000-00805f9b34fb)
	UUID: Vendor specific           (00005005-0000-1000-8000-0002ee000001)
	UUID: Message Notification Se.. (00001133-0000-1000-8000-00805f9b34fb)
	UUID: Phonebook Access Server   (0000112f-0000-1000-8000-00805f9b34fb)
	UUID: Message Access Server     (00001132-0000-1000-8000-00805f9b34fb)
	UUID: Headset                   (00001108-0000-1000-8000-00805f9b34fb)
	Modalias: usb:v1D6Bp0246d0532
	Discovering: no

`

**Comment by Poohl** [comment]
@dentedghost you need your switch's mac. since you paired successfully (did you though? `01:23:45:67:89:ab` is not a accepted MAC, it has to start with `94:58:CB:...`)

Easiest way to view it after pairing is `bluetoothctl paired-devices` it will show up as `Nintendo Switch`

**Comment by MadeSimpleWorkshop** [comment]
@Poohl Thanks Paul,  that did it. 

I was able to get my scripts working again. Even it it is hacky.

I needed to reconnect multiple times.  I got asphalt 9 working with my Bot Designer https://github.com/dentedghost/joycontrol_bot_designer which leverages https://github.com/choss/joycontrol_rest_api and https://github.com/mart1nro/joycontrol.

@thxomas and @JaredEzz Amazing job until we have a complete solution.  

I'm running the python3 simple-agent.py 

My bot been running for an hour now.  Hopefully it's easy to restart again.

**Comment by JaredEzz** [comment]
Hmm.. my bot disconnects after like 3 commands @dentedghost. What's the difference between pokemon and other games like Asphalt and Botw?

**Comment by MadeSimpleWorkshop** [comment]
@JaredEzz 

I just restarted my raspberry pi and got it to work again.

Note I needed to kill any remaining python scripts I had running
`ps -aux | grep python` this is probably a glitch on my system.

Steps:
> python3 simple-agent.py 

This is my output with multiple RequestAuth
`simple-agent.py:157: PyGIDeprecationWarning: GObject.MainLoop is deprecated; use GLib.MainLoop instead
  mainloop = GObject.MainLoop()
Agent registered
RequestAuthorization (/org/bluez/hci0/dev_70_48_F7_D4_8D_0D) authorized `

Next connect controller screen,  which like the rest of us throws an error.   

Reconnect

Note i'm using joycontrol_rest_api  example:
`curl -X 'POST' \
  'http://0.0.0.0:8000/controller/connect' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "controller_type": "PRO_CONTROLLER",
  "reconnect_address": "70:48:F7:D4:8D:0D"
}'`

Then try to move to game.  Note again.  I sometimes need to reconnect again.

In game,  I need to reconnect 3 times before I could use my script (joycontrol_bot_designer).

Can you try reconnecting multiple times to see if it just needs to stick?

**Comment by JaredEzz** [comment]
@dentedghost tried again, reconnecting multiple times in game. Each time I get this error.
```
[17:35:50] joycontrol.transport write::155 ERROR - [Errno 104] Connection reset by peer
[17:35:50] joycontrol.protocol connection_lost::117 ERROR - Connection lost.
[17:35:50] joycontrol.protocol input_report_mode_full::209 ERROR - [Errno 104] Connection reset by peer
```
This is the response I get when the switch goes into sleep mode, FWIW. Also I don't think just getting it to connect in-game will work for me since part of my script requires closing and reopening the game. Thanks for the help though!

**Comment by thxomas** [comment]
Guys, I'm in hell... I cannot have the Switch to pair anymore since yesterday :( I use the v12_fixes branch now, everything should be fine (MAC Addr., Name, SDP, simple-agent, Device Class, unpaired, /var/lib/bluetooth wiped, ...) but it is not... I'm back at the start of this thread : "User Auth accepted", then "Link Key Exchange" then stall.
I can see these events in btmon, but no connection request ever reaches joycontrol socket..
Are you still having success on your side ?

**Comment by Poohl** [comment]
@thxomas I just checked and I can get to python code just fine on the v12_fixes branch (not much further because dirty broken tree :| ).

Have you tried a different MAC? maybe the switch assumes something for yours by now?

**Comment by illie0** [comment]
> Please be aware it is still manual, and not ready for easy sharing. Steps to reproduce :
> 
> * **Change RPi BT MAC address to get in Nintendo OUI range** : I just changed the first 3 byte of the original to 94:58:CB
>   `hcitool cmd 0x3f 0x001 0x94 0x58 0xCB 0xbe 0x13 0x37`

FYI, the MAC address needs to be in reverse order with the hcitool command, should be:
hcitool cmd 0x3f 0x001 0x37 0x13 0xbe 0xcb 0x58 0x94

You can verify by using 'hcitool cmd 0x04 0x009'

**Comment by thxomas** [comment]
Wow ! Thanks Paul ! (facepalm) I'm back with you although too late for today.

**Comment by Poohl** [comment]
@illie0 what bluetooth controller are you using (manufacturer & how it's connected to PC)?

These magic numbers are very hardware-specific and in my experience the `bdaddr` tool is aware of these numbers for a few manufacturers but reversed MAC seems stange.

**Comment by illie0** [comment]
This is the onboard controller on a raspberry pi 3b. Here's showing the factory MAC, and hcitool reads it in reverse order. Setting it in reverse and re-querying it shows it still in the same reversed order as the original.

```
root@raspberrypi:~# hciconfig hci0
hci0:   Type: Primary  Bus: UART
        BD Address: B8:27:EB:54:63:F3  ACL MTU: 1021:8  SCO MTU: 64:1
        UP RUNNING PSCAN
        RX bytes:1531 acl:0 sco:0 events:97 errors:0
        TX bytes:3060 acl:0 sco:0 commands:97 errors:0

root@raspberrypi:~# hcitool cmd 0x04 0x009
< HCI Command: ogf 0x04, ocf 0x0009, plen 0
> HCI Event: 0x0e plen 10
  01 09 10 00 F3 63 54 EB 27 B8
root@raspberrypi:~# hcitool cmd 0x3f 0x001 0x37 0x13 0xbe 0xcb 0x58 0x94
< HCI Command: ogf 0x3f, ocf 0x0001, plen 6
  37 13 BE CB 58 94
> HCI Event: 0x0e plen 4
  01 01 FC 00
root@raspberrypi:~# hcitool cmd 0x04 0x009
< HCI Command: ogf 0x04, ocf 0x0009, plen 0
> HCI Event: 0x0e plen 10
  01 09 10 00 37 13 BE CB 58 94
root@raspberrypi:~#

```

**Comment by mbacalan** [comment]
Switch System Update v12.0.1 is available now, with more changes to Bluetooth it seems. ([Source](https://twitter.com/OatmealDome/status/1384320945935032323?s=20))

Sharing in case it changes anything for these fixes.

**Comment by Poohl** [comment]
@mbacalan thanks, saw it earlier too. Didn't update yet, can anyone check if that update breaks stuff even further or if it actually relaxes things? I've seen reports that the new checks actually cause joycons to sometimes also no longer pair.

**Comment by Venryx** [comment]
@Poohl I don't know about joycontrol, but for NXBT, the v12 branch still works for me (as I [first described here](https://github.com/Brikwerk/nxbt/issues/18#issuecomment-822151271)), even after updating my Switch to 12.0.1.

So "a pathway is still open" at least. (though I was never able to get the v12 joycontrol branch working, so it's possible there are implementation details between joycontrol and NXBT which could be/become relevant)

**Comment by Poohl** [comment]
@Venryx Do you happen to have a guess why joycontrol didn't work? Never connected/connection timed out? Connected but didn't pair, so the reconnect trick didn't work? 

But thanks for the input on changes, so the update is seemingly not too bad.

I read through all of NXBT's V12 changes and all they do is limit the send-rate (which doesn't work in Joycontrol for some reason, asyncio-scheduler still hat a different plan), automate what we still do manually (MAC and SPD), and some esotric additional sockets that seem to never be used.

**Comment by Venryx** [comment]
> Do you happen to have a guess why joycontrol didn't work? Never connected/connection timed out? Connected but didn't pair, so the reconnect trick didn't work?

I would get to the step where I'm supposed to wait for the Pair popup to show, but it never showed up (even with the reconnect trick). I unfortunately do not remember much more details than that.

Note that it's possible I did not follow the joycontrol instructions properly; I had already gotten NXBT working, so my motivation to thoroughly follow every step for joycontrol was less than it could have been (also, the instructions for the v12 branch were not entirely clear imo -- I had to read through the comment threads so may have missed something, for example maybe I needed to do slightly different commands for the SPD step given that my BT adapter might be different).

**Comment by Venryx** [comment]
By the way, do you have any idea why with NXBT I am only able to have one joycon running alongside NXBT's simulated controller? When I add two joycons alongside, the simulated controller disconnects every time as soon as I leave the change-grip-order panel. (I can add a second physical controller if I use the Magic-NS adapter however [eg. for my Xbox controller], presumably because it uses the USB port rather than Bluetooth, and/or because it avoids having both a left and right joycon connected.)

Maybe this could be a hint as to what properties are not yet being fully simulated correctly by NXBT and joycontrol?

**Comment by Poohl** [comment]
@Venryx I know of that problem, NXBT is aware. I'm assuming something to do with either
- rate limiting on the switches end, 
- something with the timing going out of whack
- or bluetooth adapters intended for host operation producing too many collisions.

**Comment by Venryx** [comment]
>  I know of that problem, NXBT is aware.

So this issue (of there being a limit on physical controllers when a simulated controller is used) was present pre-12.0.0?

Anyway, I might try fiddling with the BT data being sent out to see if I notice any stability effects on that front... (would be nice to be able to use the full set of physical controllers even when NXBT, or joycontrol, is active)

**Comment by Poohl** [comment]
@Venryx I don't know, sorry.

**Comment by MadeSimpleWorkshop** [comment]
I’m still partially working even with the newest 12.01 update.  Ha finally disabled after that one

**Comment by thxomas** [comment]
@Poohl 
I've been looking at some wireshark traces, and have a theory that deserves a test ;)
I don't have time now, but if someone wants to try issuing `hciconfig hci0 lp RSWITCH` before trying to reconnect... It might improve the connection stability by disabling SNIFF mode policy. (I noticed Mode Switch to Sniff Mode then back from the controller just before disconnection in game., The console may not expect it, thus dropping the connection)

Let me know !
👍 

![image](https://user-images.githubusercontent.com/8565794/115603390-26667900-a2f1-11eb-8e59-ebd4a94c46f9.png)

**Comment by Poohl** [comment]
@thxomas If that is a thing thing it might dislike I'll try. I tried to hook into the HCI_EVT called "Max Slots Changed" to a value of 1 (from 5) yesterday:
```
041b030b0001
```
This event happens reliably both on axed and on working connections. It also happens when relaying a joycon, although they seem to not care about any of this, as they just always send with 60 Hz somehow.

My understanding of this is that now there is more bandwidth available as the channel is now shared among 1 instead of 5 devices? Has anyone ever tried to pair 6+ controllers on the Change grip screen? did it work?

I tried to increase the frequency when this event fired to no avail. If anyone else has suggestions?

**Comment by Poohl** [comment]
also @thxomas (or anyone intrested): if you tell me your version of wireshark I can send you my [nxbt-dissector](https://github.com/Poohl/Nintendo_Switch_Reverse_Engineering/tree/nxbt_wireshark_plugin/packet_parse/nxbt) plugin already compiled. to see useful information about the switches talk.

**Comment by thxomas** [comment]
Hi Paul, I'm running wireshark v3.4.3

**Comment by Poohl** [comment]
I pushed the last changes to the repo above if you want to check out the code, otherwise here is the binary (compiled on ubuntu 20.04 x64, hope that works): https://drive.google.com/file/d/1DnH_IsVTXK8BvM_g1wrAQfrreWdRKLRt/view?usp=sharing

copy the file to `~/.local/lib/wireshark/plugins/3.4/epan/` and restart wireshark. I reccommend you start wireshark from the terminal, then you should see a message that the dissector was registered (and any errors it produced during dissection).

Note that restarting is in general a good idea with this code. It's not perfect.

**Comment by thxomas** [comment]
OK, I should have mentioned I remotely wireshark on windows (but plugins are scripts aren't they ?)

BTW, I've got a second setup : freshly installed Pi4. Paired at the first attempt, not disconnecting for minutes on BotW. No Sniff Mode changes until now.
(Testing with both joycons detached and off)

I will focus on nfc debugging, I get disconnected as soon as I send 'zl'

**Comment by Poohl** [comment]
This plugin is compiled C code sadly. I dunno how to translate that to Windows yet (i don't even know how plugins work on windows). I might port this to LUA some time, but I just found another bug wit the MCU responses all beeing unknown anyway...

**Comment by Poohl** [comment]
@pokemon-bot Yeah, also discovered similar weirdness. It seems whatever was changed on the switches end wasn't to detect this emulation (or it's really bad at it). If you don't mind, try Wiresharking this kind of stuff as there has to some difference between stable and aborted connections. We just have to find it...

Just for the record, stuff that seems to work:
- wrong MAC address
- wrong Name, but correct Alias
- input reports at 15-30Hz on the pairing screen
- sometimes getting out of the pairing screen before blueZ blocks due to lag
- sometimes getting out of the pairing screen before it's fully paired?
- Seconds worth of lag (tested up to 15s)
- just not sending anything for minutes on end on the pairing screen ???
- Wrong length input reports (too short on x30, to long on x3F)

What doesn't work:
- MAC's that previously messed up during pairing (even when unpaired)
- Names that previously messed up during pairing (even when unpaired)
although I don't know if this is actually only because fixing these things resets parts of BlueZ

At this point I believe there is either very sloppy rate-detection going on or It's something low level on the Bluetooth controller that causes a headache (maybe dropping packets?).

**Comment by damos1212** [comment]
Controller input + amiibo reading on 12.0.1 worked for me using the V12_fixes branch on a ubuntu vm using thomasx's information.

**Comment by Poohl** [comment]
@damos1212 WHAT? Without any modifications to the NFC-code?

What game, controllertype and bluetooth chipset are you using?

Thxomas said he couldn't get NFC to work and my two attempts (on 12.0.0) were also futile.

**Comment by damos1212** [comment]
I tried the magnamalo amiibo in monster hunter rise. Used pro controller arg and am using a ASUS USB-BT400. This was all done in ubuntu 21.04. At first it just kept disconnecting in the amiibo screen but somehow I managed to get it to work by reconnecting and immediatly loading the amiibo. I'll see If I can reproduce it.
Can't get it to work again so far. All I know is that it just stopped disconnecting on the amiibo screen and displayed a different number from joycontrol.transport read::77 DEBUG after a bajillion tries and a switch crash.

**Comment by mbacalan** [comment]
I also managed to get a stable connection once with V12_fixes branch and Switch v12.0.1. Though after accidentally closing the connection while trying to try NFC, I couldn't get a reconnection. Will try more later when I have time.

Edit: I can get it to connect reliably using the `-r` argument and having 1 joycon connected. Only works if you're on the Switch menu and not in a game. I can add a `.bin` file with `nfc` command but when I try to use Amiibo functionality in a game, it errors and disconnects. Tried with ACNH and Zelda BotW.

```
cmd >> [16:58:04] joycontrol.transport read::81 ERROR - No data received.
[16:58:04] joycontrol.protocol connection_lost::124 ERROR - Connection lost.
[16:58:04] asyncio default_exception_handler::1707 ERROR - Task exception was never retrieved
future: <Task finished name='Task-33' coro=<L2CAP_Transport.close() done, defined at /home/mb/Desktop/poohl-joycontrol_v12/joycontrol/transport.py:148> exception=AttributeError("'NoneType' object has no attribute 'cancel'")>
Traceback (most recent call last):
  File "/home/mb/Desktop/poohl-joycontrol_v12/joycontrol/transport.py", line 155, in close
    if self._read_thread.cancel():
AttributeError: 'NoneType' object has no attribute 'cancel'
```

**Comment by Poohl** [comment]
@pokemon-bot you need to do one obvious and one tedious thing:
- hook into `connection_lost` in `protocol.py`. This function (should, if not check `transport.py`) be called each time the connection dies. To use that signal I think it's best to only pass it along to `command_line_interface.py` or handle it within the protocol itself.
- Make sure all the asyncio tasks restart. There are a few started with `asyncio.ensure_future` handling in and output. I know of 2: the reader in transport.py (the one that always dies on disconnect) and the writer in protocoll.py.

**Comment by Poohl** [comment]
@pokemon-bot I'm pretty sure there's a way, although I don't really know how the API reacts to the internals crashing. Have you tried just restarting if e.g. no input was sent for a second?

**Comment by omarasahmad** [comment]
Whenever I try and do the -r command (with all the right information required) I keep getting a "command not found" message. I tried reinstalling Joycontrol but it did nothing. Am I missing something?

**Comment by Poohl** [comment]
Ok, I thought a bit about alternitives to getting the information we need to figurt out what the switch is expecting. As far as I can tell, there are mainly 4 options:
* hooking into the actual BT-communication by hooking into the probe-points in the switch. From googling and [this repo](https://github.com/hedgeberg/Switch-Logic-Captures) i found it's [I3 and I4](https://switchbrew.org/wiki/Testpads#Cluster_I) that carry the Bluetooth commands to and from the controller. Once captured these can be fed into Wireshark and then the NXBT dissector to see what's going on. Doing this on the coltroller side seems to be impossible, as the controller is in the same chip as the BT-chipset.
* hooking into the software-stack on a softmodded switch. There are APIs for bluetooth available and thes seem to be well understood?
* diassembling the necessary firmware parts. For the switch some efforts seem to be made as [switchbews changelog](https://switchbrew.org/wiki/12.0.1) mentiones changes in the implementation of various bluetooth things. For the Joycons the main problem would be getting any useful specsheet.
* Using an emulator/sandbox to simulate the given system's behaviour. This seems like a far fetch, but if someone has a single-steppable emulator lying around?

**Comment by Poohl** [comment]
@TrolledOutGamer the `-r` is a start-parameter not a command. Depending on how you start joycontrol, you will have to supply it at startup:
```
sudo python3 run_controller_cli.py -r AA:BB:CC:DD:EE:FF <controller_type>
```

**Comment by omarasahmad** [comment]
Thank you! Unfortunately it still auto-disconnects :( new switch update threw

**Comment by Poohl** [comment]
Hey, just found the dumbest thing alive:
If you press "up" 4 times on the pairing screen you can leave it no problem.
To reporoduce:
```
./scripts/unpair.sh
sudo python3 run_controller_cli.py PRO_CONTROLLER
<wait for the subcommands to finish>
click up
click up
click up
click up
click a
```
I have done this 3 times by now and it worked every time.
Can anyone confirm this? NXBT somewhere mentioned having to press L and R to not get kicked. Is this the same weird fix for the same weird issue?

But still no idea why on NFC you get a instant disconnect. Currently fiddeling with HCI messages.

**Comment by thxomas** [comment]
I'm stuck too at finding any clues on the NFC-instant disconnect.

So far, my most stable setup is with RPI4. I can pair everytime, and once paired, I reconnect OK.
Reconnection works only in Home Menu, if I reconnect ingame, it gets disconnected instantly, but if I go/launch game from Home, the connection is stable.

HCI messages on disconnect seem to indicate that the Switch has decided to disconnect. The sequence is always the same and I'm trying to understand if the mode switch to Sniff is the cause or consequence of the disconnection.

> ![image](https://user-images.githubusercontent.com/8565794/115603390-26667900-a2f1-11eb-8e59-ebd4a94c46f9.png)

Maybe the best way to go further now, is to man-in-the-middle btw a true controller and the switch, could you give some additional hints on using joycon_ip_proxy.py ? would it run with PRO_CONTROLLER too ?

At the software level, joycontrol does not even receive the first NFC command.

**Comment by Poohl** [comment]
@thxomas I can give you wireshark captures for that if you want, I'll also add some more doc for `joycon_ip_proxy.py`

But don't get your hopes up too high, limitations not really of the script but more of the nature of what it tries to do:
- LAG, and often a lot of it. Despite messages beeing in flight for just about 1ms longer the lack of flow control causes bad stuttering when approacing the limits of the BT-controller (thats why I implemented the IP part, didn't help too much).
- bluetooth is not made for chains, only for point to point. A lot of bandwidth is lost due to the controller constatntly switching between Host and Slave mode, toggeling sniff mode etc...  
  Read this as PRO_CONTROLLER is basically impossible due to the need for >250 Messages/sec and anything NFC also only works for 1-2 seconds before the switch throws some kind of error and reverts back to 0x30 mode.
- ~I didn't try since the V12 update, because every time you do this it messes up the pairing and until yesterday I barely got that to work, but I would assume the timing is no better with the proxy due to the lag.~
Edit: scap that last one, I have a sucessfull pairing capture somehow. See next comment.

**Comment by Poohl** [comment]
To help the effort and because the proxy stuff is really spotty, I uploaded my captures for analysis by whoever is interested to [Gdrive](https://drive.google.com/drive/folders/1ce2bmYrhA-dMYxEW4sOES4MnUq52oz-h?usp=sharing).
Grouped in real, that is proxied, and simulated ones. Hope the titles are descriptive enough.
~I'll also push another one captured by @yamakay of the Switches UART-HCI-Bus later. But it's already older, more as a proof of concept that it can in fact be done and going this route can yield results.~
Edit: The UART-Captures were by @hedgeberg from his [Switch-Captures](https://github.com/hedgeberg/Switch-Logic-Captures) repo and parsed using info from [switchbrew](https://switchbrew.org/wiki/Testpads#Cluster_I).
@Yamakaky stuff is coming too, for anyone wanting to do the same: I use a version of wireshark with [this merge request](https://gitlab.com/wireshark/wireshark/-/merge_requests/2136) merged (currently only the master branch) and import from [hexdump using regex](https://www.wireshark.org/docs/wsug_html_chunked/ChIOImportSection.html) to convert whatever-dumps to pcapng.
When you have the raw HID-Communication, choose encapsulation "USER_0" or  "Upper PDU" and "nxbt" and you can then easily dissect it using [my dissector](https://github.com/Poohl/Nintendo_Switch_Reverse_Engineering/tree/nxbt_wireshark_plugin/packet_parse/nxbt)

**Comment by Yamakaky** [comment]
It's @Yamakaky ;)

**Comment by Yamakaky** [comment]
Did you manage to get your `proxy.py` to work?

**Comment by Poohl** [comment]
@Yamakaky yes.
It's the [`joycon_ip_proxy.py`](https://github.com/Poohl/joycontrol/blob/V12_fixes/scripts/joycon_ip_proxy.py) mentioned above.
I updated the doc in there (top of the file) to hopefully clear up how it works.

Pre V12 I've had best results when reconnecting, Since then I couldn't get that to work yet because something always didn't cooperate and I just got to frustrated after a while of switch or joycon disconnecting for no good reason.
But pairing has worked on V12 already, so I bet that's just a matter of it working once...

**Comment by Yamakaky** [comment]
Oh right, I was checking in nxbt. I finally managed to do the pairing with this script, but key presses were delayed a lot then it disconnected. When reconnecting, the forwarding starts but the joycon turns off at the first keypress. I guess buying a usb bluetooth adapter is unavoidable ?

**Comment by Poohl** [comment]
I don't really know what the problem is to be honest. I tried to use 2 different bluetooth controllers (hence the IP-part, don't have a USB BT dongle) but that had lots of lag always and didn't really improve anything else.
A tip for getting it out of the pairing screen is to just be fast... the lag builds up and as long as it's not too bad you can get out of there.

**Comment by Yamakaky** [comment]
Hum, in the current state there is no buffering, so the script will not receive messages until the last one was sent. Maybe using a bounded queue could help?

**Comment by Poohl** [comment]
@Yamakaky could be. Someone on stackoverflow also suggested using some asyncio-wrapper around the sockets, presumably with buffering. I assumed linux buffers sockets anyway and the lag caused by two syscalls to a buffer is neglectable compared to python. So i never tried anything like that, might be worth a shot.

**Comment by Yamakaky** [comment]
I did that on https://github.com/Yamakaky/joycontrol/tree/V12_fixes. One strange thing I notice is that after clicking on ZR, ZL or ZR, a lot of messages are sent (probably the 66Hz input report). After a few seconds, it goes into a loop where everything works fine for a few seconds then messages from the joycon gets buffered for a few seconds. Could you try with two adapters?

**Comment by Poohl** [comment]
@Yamakaky I'm sorry, i cant get BlueZ to cooperate. It won't connect to the friggin _joycon_. at all. seems to have turned on it's error-message randomizer again. Will sacrifice a few of their devs to Cthulhu maybe it works then. F***ing garbage software.

**Comment by Yamakaky** [comment]
BTW I can't manage to reconnect the devices, I have to remove and pair them each time.

**Comment by Poohl** [comment]
@Yamakaky The good news: I got it working after presumably 4 BlueZ devs. I had to convert it to UDP instead of TCP, because L2CAP is frame based and when using queues+asyncio even without nagle's algorithm linux starts combining frames in TCP. That also improved lag and stability quite a bit, since UDP is generally more similar to L2CAP. (will push soon)
However I _only_ got it working with the reconnect option, as the pairing is just way to spotty for me. This wasn't really a problem, the datarate is really managable (I tried with 2 bt-controllers, one on a raspi, one on my Desktop. ping between both machines is 0.2ms).
I tried going around the Homescreen and started untitled Goose game, all without issue (and without Max Slots Changes).
Now the very bad news: After the max slots change event it just instantly kicks the proxy just like the simulated controllers (i tried to read an amiibo in the settings). However it does not reliably trigger that event when connecting a new controller. (it worked once and kicked me twice).
This means whatever causes the disconnect is either a) not propagated by the proxy or b) already has too much lag. Considering the shitshow that is this protocol's timing and the fact that I captured the end closer to the switch I'm heavily siding for a). While the SR, SL and ZR buttons seem to indicate something on the Pairing screen I can't find anything in the trace I got now that would indicate any change at the ACL layer or higher. Meaning something is going on in the lower Bluetooth levels. And the only things lower than ACL is HCI, H4 and the Air...
In the process I also noticed one more thing: You can send completely malformed input reports as long as they are valid L2cap (start with `a1`) the switch will keep the connection.

**Comment by Poohl** [comment]
Ok Great news: the Grand Patch is live, please everyone try the following:

```
sudo run_controller_cli.pi PRO_CONTROLLER
(read/react to prompts if you get any, feedback is appreciated)
(wait for the 4 print messages)
click up
pause
click a (to leave the pairing screen)
unpause
```

And to read amiibos:
```
nfc <amiibo_dump.bin> (or use the --nfc argument)
pause
click <whatever it takes to start the reading, usually A>
unpause
```
Try the same for writing amiibos. Untested yet. It is important, that whenever you want to enter any amiibo-screen you are paused.

In general that is the main new thing. When you are paused the switch doesn't axe the connection (at least for a while), so whenever you come to some moment where you assume it's about to happen, try to pause. Note however that anything using the gyro this might not work or result in eradic movement. Also while paused anything but the buttons don't work.
Any feedback regarding success/faliure is appreciated. Most notably any pauses you had to insert into whatever you are usually doing would be appreciated to figure out what and where it needs automating.

Also another debug feature is in there:
```
debug X
```
sets the input frequency to XHz, if you get a lot of `running too slow` warnings.

**Comment by mbacalan** [comment]
I had to give it a few tries and follow the instructions on #4, even though it felt a little random as to why it connected then and why it didn't before, I was able to move around in Animal Crossing and use an NFC too! I did get a disconnection after using the NFC but unlike before, it did read the NFC before disconnecting and I could reconnect right away using the `-r` parameter.

So this looks like a great work to me. Thanks a lot @Poohl !

One question, I do get a lot of `running too slow` warnings but in what range should I try using `debux X` command?

**Comment by mbacalan** [comment]
BTW, tried it with Zelda BotW too and NFC worked once but can't get it to work again, tried 4 times with disconnections. Game says use a controller that supports NFC or not have the Pro Controller charging.

For reconnecting, I have to close the game or it will error immediately.

**Comment by Poohl** [comment]
@mbacalan officially the `debug X` argument should be 60 for joycons or 120 for the pro controller, but that sometimes doesn't work. Still don't know why. Sometimes you have to go as low as 4 to get rid of the errors.
And just as a warning: anything below 1 works, but makes everything really unresponsive.
Also just to clarify: you got disconnects while paused? Instantly as before or after a few seconds beeing paused? Try waiting a bit before starting the read, it takes a while until the pause takes effekt on the hardware. (You can open Wireshark and see when the packets stop)
And for the weird "not an nfc controller" problem please see #5. Still no idea there.

**Comment by mbacalan** [comment]
@Poohl yes, I have tried again now with longer times after pausing. It did work once again but couldn't get it to work again even after rebooting both devices and trying multiple times. So that seems to be a little random, at least with BotW. ACNH worked on first try, not sure if it was coincidence or not.

The disconnect happens while paused / without using `unpause`.

**Comment by Yamakaky** [comment]
On my side, the pairing works, but the connection is lost as soon as I leave the pairing screen:
```
[17:40:41] root _reply_to_sub_command::191 INFO - received Sub command SubCommand.ENABLE_6AXIS_SENSOR
[17:40:41] root _reply_to_sub_command::191 INFO - received Sub command SubCommand.ENABLE_VIBRATION
[17:40:41] root _reply_to_sub_command::191 INFO - received Sub command SubCommand.SET_NFC_IR_MCU_CONFIG
[17:40:41] root _reply_to_sub_command::191 INFO - received Sub command SubCommand.SET_PLAYER_LIGHTS
This shuld be false: True
[17:40:41] joycontrol.protocol _writer::151 INFO - writer started
[17:40:41] joycontrol.protocol _pair_inputter::130 INFO - pairing writer started
shouldnt be here
shouldnt be here
shouldnt be here
shouldnt be here
click up
cmd >> pause
cmd >> click a
cmd >> [17:40:59] joycontrol.transport read::118 ERROR - No data received.
[17:40:59] joycontrol.protocol connection_lost::249 ERROR - Connection lost.
[17:40:59] asyncio default_exception_handler::1738 ERROR - Task exception was never retrieved
future: <Task finished name='Task-246' coro=<L2CAP_Transport.close() done, defined at /home/mikael/dev/python/joycontrol/joycontrol/transport.py:189> exception=AttributeError("'NoneType' object has no attribute 'cancel'")>
Traceback (most recent call last):
  File "/home/mikael/dev/python/joycontrol/joycontrol/transport.py", line 196, in close
    if self._read_thread.cancel():
AttributeError: 'NoneType' object has no attribute 'cancel'
```

**Comment by mbacalan** [comment]
@Yamakaky this happened to me once on initial pairing (opening the controllers menu) but after that, I'm able to connect using the `-r MAC_ADDR` flag.

**Comment by Poohl** [comment]
Even more amazing news: It's now fully automated. I got the lag in Linux under control and could implement a transport-level fix.

Why transport level is good: no game specific detection necessary, no input, output or whatnot dependencies to check. No need to handle NFC-reading, NFC-writing, Pairing etc... Let's hope that's really it.
Only caveat: you'll get a warning that the code is running very slow every time this fix kicks in.

**Comment by mbacalan** [comment]
@Poohl latest version is working perfectly for me, amazing work! Connecting with `-r` flag never fails and I used over 10 NFC reading with no disconnects or any other issues.

**Comment by dtrunk90** [comment]
Hey guys. You don't have to mess with Mac addresses. Changing the device class is enough to be able to connect again.

**Comment by thxomas** [comment]
@Poohl wow, GG on the great progress ! 

Just tested the Grand V12 Fix on RPi4 : both joycons attached, pairing at 1st attempt and still on the initial connection 25 min later ! After exiting Grip menu, I started ACNH, read an amiibo card, back home, managed to write to a card, swapped to BOTW, runs OK. goes to controller screen if I use the joycons, and finally crashed some way. After easy reconnect, amiibo use is OK ! 

Some rare "Code too slow" (a burst when reconnecting while ingame)

This issue is gonna close soon... 👏 
@thxomas

**Comment by Poohl** [comment]
@thxomas Yes, am doing a bit of polish and final testing currently. Just read, wrote and re-read 3 amiibos. Paired and reconnected no problem. Went in and out of game, etc.
The "please press R+L on the first controller screen" always has been a problem, no clue how the joycons salvage that one.
Plan of action on my end:
- touchup the transport code (the `_flow_control_???` stuff is a huge mess)
- touchup the server (suppress questions to user when used as library)
- remove dead code from protocoll (there is an `anticipate` command in there that doesn't work)
- merge everything back into `amiibo_edits`.
- Close this and the NFC issue
- update pull requests

Still TODO:
- go into 60hz mode when leaving the pairing screen
- clean shutdown of the dozens of aio tasks
- update documentation, i added an `-r auto` option
- remove the `pause` command. They act on protocoll level and might still be usefull.

**Comment by Poohl** [comment]
Ok, I can pair and reconnect, write amiibos etc no problem.

For now I consider this closed.

**Comment by Venryx** [comment]
@Poohl Are the instructions in this repo's readme still sufficient for v12.0.1?

Or are there additional instructions (or no-longer-needed instructions) that still need to be merged?

**Comment by Poohl** [comment]
@Venryx The instructions in the readme should still work, but joycontrol will now also try to fix these things itself or warn you (wt least when using the cli).
I'll update the readme tomorrow.
The only thing notably different to the doc are the dependencies. you need
- `hcitool` or `bdaddr`
- `hciconfig`
- `systemctl`

Whether any of these changes are no longer needed only further testing will tell. @dtrunk90 stated that the MAC-stuff is uneccessary. Someone else reported success with the old Plugin-setup. I had success without the Name change.

It might be the old steps are still all you need and everything else is just panic-fixes for nonexistent problems. Will test a bit the next week.

**Comment by munichi-jasuda** [comment]
I can now successfully connect to my Switch using the latest version by Poohl, but can't read Amiibo. Connection closes when attempting to read Amiibo. (I have also changed the ExecStart parameters in bluetooth.service, bluetooth MAC address using the change_btaddr.sh, and bluetoothctl system-alias).

![Error](https://user-images.githubusercontent.com/84059601/117917761-6a083d80-b307-11eb-8dab-84150ea9f2b2.jpeg)

**Comment by thxomas** [comment]
Ouch, v12.0.2 is available... misclick, big sweat.... but everything you fixed seem to work in this new version.

**Comment by munichi-jasuda** [comment]
The v12.0.2 makes changes to Bluetooth Driver. Is it safe to update or should I stick to v12.0.1?

**Comment by Poohl** [comment]
@munichi-jasuda 
> Connection closes when attempting to read Amiibo.
That looks like #5, and we still don't really know what the issue is.
You too can try to start `tshark` before joycontrol and send me the capture. I currently can't reproduce it so this is the best i can offer regarding debugging.

**Comment by munichi-jasuda** [comment]
I think the problem is definitely with the device class. Even after changing the MAC address and successfully connecting with the Switch, the device is eventually identified as invalid when reading NFC.

I ran the `bluetoothctl show` command both before and after successfully connecting with my Switch, and it seems my device class is unaffected (**0x001c0000**). It's not **0x2508**, as it should be in order to be identified as a valid controller that can read Amiibo.

**Comment by dtrunk90** [comment]
Mac address changing can only be considered as a workaround as you can Set the Mac address to one of your Real controllers who already got paired successfully to the switch. If you then connect with a Mac address of one of your Real controllers switch Firmware doesn't check the device class for pairing. So this is only bypassing the device class check when pairing. The Real solution is only to change the device class. That's what we already found out with Joycon Droid.

**Comment by Poohl** [comment]
@dtrunk90 So even the SDP-changes are unnecessary? Meaning the switch will pair despite too many capabilities, a non-nintendo mac and a random name and alias, as long as the class is ok? or is the Alias still required?

**Comment by dtrunk90** [comment]
> @dtrunk90 So even the SDP-changes are unnecessary? Meaning the switch will pair despite too many capabilities, a non-nintendo mac and a random name and alias, as long as the class is ok? or is the Alias still required?

I can only speak for pairing because i dont have the joycondroid Source Code. But yes, for pairing only the device class is needed. Nothing else. We made a magisk module in order to be able to keep using joycondroid.

**Comment by Poohl** [comment]
@dtrunk90 Thanks for the info, i updated #4 and pushed a commit to `server.py` to no longer mess with the mac.

**Comment by thxomas** [comment]
I tested too : Changing the MAC is not mandatory after all.
